sdk/testing/stepwise/stepwise.go

```diff
@@ -0,0 +1,351 @@
+			}
+		}
+		tt.Fatal(driverErr)
+	}
+
+	defer func() {
+		if c.SkipTeardown {
+			logger.Info("driver Teardown skipped")
+			return
+		}
+		if err := c.Environment.Teardown(); err != nil {
+			logger.Error("error in driver teardown:", "error", err)
+		}
+	}()
+
+	// retrieve the root client from the Environment. If this returns an error, fail
+	// immediately
+	rootClient, err := c.Environment.Client()
+	if err != nil {
+		tt.Fatal(err)
+	}
+
+	// Trap the rootToken so that we can preform revocation or other tasks in the
+	// event any steps remove the token during testing.
+	rootToken := c.Environment.RootToken()
+
+	// Defer revocation of any secrets created. We intentionally enclose the
+	// responses slice so in the event of a fatal error during test evaluation, we
+	// are still able to revoke any leases/secrets created
+	var responses []*api.Secret
+	defer func() {
+		// restore root token for admin tasks
+		rootClient.SetToken(rootToken)
+		// failedRevokes tracks any errors we get when attempting to revoke a lease
+		// to log to users at the end of the test.
+		var failedRevokes []*api.Secret
+		for _, secret := range responses {
+			if secret.LeaseID == "" {
+				continue
+			}
+
+			if err := rootClient.Sys().Revoke(secret.LeaseID); err != nil {
+				tt.Error(fmt.Errorf("error revoking lease: %w", err))
+				failedRevokes = append(failedRevokes, secret)
+				continue
+			}
+		}
+
+		// If we have any failed revokes, log it.
+		if len(failedRevokes) > 0 {
+			for _, s := range failedRevokes {
+				tt.Error(fmt.Sprintf(
+					"WARNING: Revoking the following secret failed. It may\n"+
+						"still exist. Please verify:\n\n%#v",
+					s))
+			}
+		}
+	}()
+
+	stepCount := len(c.Steps)
+	for i, step := range c.Steps {
+		// range is zero based, so add 1 for a human friendly output of steps.
+		// "index" here is only used for logging / output, and not to reference the
+		// step in the slice of steps.
+		index := i + 1
+		if logger.IsWarn() {
+			progress := fmt.Sprintf("%d/%d", index, stepCount)
+			logger.Warn("Executing test step", "step_number", progress)
+		}
+
+		// reset token in case it was cleared
+		client, err := rootClient.Clone()
+		if err != nil {
+			tt.Fatal(err)
+		}
+
+		// TODO: support creating tokens with policies listed in each Step
+		client.SetToken(rootToken)
+
+		resp, respErr := makeRequest(tt, c.Environment, step)
+
+		if resp != nil {
+			responses = append(responses, resp)
+		}
+
+		// If a step returned an unexpected error, fail the entire test immediately
+		if respErr != nil && !step.ExpectError {
+			tt.Fatal(fmt.Errorf("unexpected error in step %d: %w", index, respErr))
+		}
+
+		// run the associated StepCheckFunc, if any. If an error was expected it is
+		// sent to the Check function to validate.
+		if step.Check != nil {
+			if err := step.Check(resp, respErr); err != nil {
+				tt.Error(fmt.Errorf("failed step %d: %w", index, err))
```

Comment:
Would some sort of name on the step be valuable here? I'm concerned about the index being misinterpreted as off-by-one when debugging.

Reply:
An optional `name` could be useful here, yes. I'm going to defer the addition of `name` for now. I've updated the documentation on `case.Steps` to reflect the 1-based indexing used in logging in 72e24e7a8